src/S3/S3Transfer/MultipartUploader.php

```diff
@@ -0,0 +1,289 @@
+                $uploadPartCommandArgs
+            );
+            $commands[] = $command;
+            $this->calculatedObjectSize += $partBody->getSize();
+            if ($partNo > self::PART_MAX_NUM) {
+                return Create::rejectionFor(
+                    "The max number of parts has been exceeded. " .
+                    "Max = " . self::PART_MAX_NUM
+                );
+            }
+
+            if ($partNo > $partsCount) {
+                return Create::rejectionFor(
+                    "The current part `$partNo` is over the expected number of parts `$partsCount`"
+                );
+            }
+        }
+
+        if ($hashBody) {
+            $this->requestChecksum = hash_final($this->hashContext);
+        }
+
+        return (new CommandPool(
+            $this->s3Client,
+            $commands,
+            [
+                'concurrency' => $this->config['concurrency'],
+                'fulfilled'   => function (ResultInterface $result, $index)
+                use ($commands) {
+                    $command = $commands[$index];
+                    $this->collectPart(
+                        $result,
+                        $command
+                    );
+                    // Part Upload Completed Event
+                    $this->partCompleted(
+                        $command['ContentLength'],
+                        $command->toArray()
+                    );
+                },
+                'rejected'     => function (Throwable $e) {
+                    $this->partFailed($e);
+
+                    throw $e;
+                }
+            ]
+        ))->promise();
+    }
+
+    /**
+     * @return int
+     */
+    protected function getTotalSize(): int
+    {
+        if ($this->calculatedObjectSize > 0) {
+            return $this->calculatedObjectSize;
+        }
+
+        return $this->body->getSize();
+    }
+
+    /**
+     * @param ResultInterface $result
+     *
+     * @return UploadResult
+     */
+    protected function createResponse(ResultInterface $result): UploadResult
+    {
+        return new UploadResult(
+            $result->toArray()
+        );
+    }
+
+    /**
+     * @param StreamInterface $stream
+     * @param array $data
+     *
+     * @return StreamInterface
+     */
+    private function decorateWithHashes(
+        StreamInterface $stream,
+        array &$data
+    ): StreamInterface
+    {
+        // Decorate source with a hashing stream
+        $hash = new PhpHash('sha256');
+        return new HashingStream($stream, $hash, function ($result) use (&$data) {
+            $data['ContentSHA256'] = bin2hex($result);
+        });
+    }
+
+    /**
+     * Filters a provided checksum if one was provided.
+     *
+     * @param array $requestArgs
+     *
+     * @return string|null
+     */
+    private static function filterChecksum(array $requestArgs):? string
+    {
+        foreach (self::$supportedAlgorithms as $algorithm) {
```

Comment:
This is fine on such a small array size, but you could make the supported algorithms array associative where the algorithm name is the key and do an isset check here